[Add/edit observation] Major cleanup of view model and view

gino-m opened the pull request:
Fixes #242.

gino-m commented:
@shobhitagarwal1612, @scolsen, some insights/reminders that came up during this exercise:

  * Views invoke methods on ViewModels directly, observe changes via LiveData. 
    * Prefer LiveData over Android and RxJava Observables, since they're lifecycle aware and
have an immutable interface.
    * Exception: Data binding can safely use Android Observables since it managed lifecycle.
   * ViewModels invoke methods on DataRepository, observe results and changes via Rx streams.
   * Side effects and statefulness are ok inside for ViewModels. Use them as an alternative to
pure streams when streams would need to be split multiple times and make code unmanageable.
   * For async operations triggered via user interaction, use a single ViewModel-scoped stream and single subscribe when creating View. Use switchMap or flatMap to run async operations so that subscriptions of repeated operations are managed by Rx.
   * For `MutableLiveData` fields, use accessor methods to expose immutable `LiveData` instead of making fields public. Theres no immutable wrapper for Android Observables, but we expose those via getter anyway for consistency.

gino-m commented:
I tried to split out some changes into smaller commits, but the bulk of the work in f3a5312efdf598aa96e67d26b09227238ea235ca was difficult to retroactively disentangle. 

@scolsen, @shobhitagarwal1612, @deepakbansal please assign yourself as Assignee if you can take a look. Thanks in advance!

shobhitagarwal1612 commented on gnd/src/main/java/com/google/android/gnd/ui/editobservation/EditObservationFragment.java:
```diff
@@ -92,67 +85,40 @@ public void onViewCreated(
     ((MainActivity) getActivity()).setActionBar(toolbar, R.drawable.ic_close_black_24dp);
+    // Initial state.
+    toolbar.setTitle("");
+    toolbar.setSubtitle("");
```
Can we use [`startWith`](http://reactivex.io/documentation/operators/startwith.html) Instead of manually setting the value to ""?

gino-m commented on gnd/src/main/java/com/google/android/gnd/ui/editobservation/EditObservationFragment.java:
```diff
@@ -92,67 +85,40 @@ public void onViewCreated(
-        EphemeralPopups.showError(getContext());
-        navigator.navigateUp();
+      default:
+        Log.e(TAG, "Unknown save result type: " + saveResult);
```
Unchecked exceptions would crash the application, which I'm not sure is what we'd want to do here.

gino-m replied:
Please reopen if necessary.

gino-m commented on gnd/src/main/java/com/google/android/gnd/ui/editobservation/EditObservationFragment.java:
```diff
@@ -92,67 +85,40 @@ public void onViewCreated(
     ((MainActivity) getActivity()).setActionBar(toolbar, R.drawable.ic_close_black_24dp);
+    // Initial state.
+    toolbar.setTitle("");
+    toolbar.setSubtitle("");
```
These are initializing the view, not the stream so that the toolbar behaves correctly before the stream emits a value. This should probably be done in the view, though.

gino-m commented on gnd/src/main/java/com/google/android/gnd/ui/editobservation/EditObservationViewModel.java:
```diff
@@ -82,89 +121,41 @@
+
+    loadingSpinnerVisibility.setValue(View.GONE);
+    saveButtonVisibility.setValue(View.GONE);
+    savingProgressVisibility.setValue(View.GONE);
```
Cool, didn't realize. Sending a commit now.

shobhitagarwal1612 commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/BaseDao.java:
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
This class is already checked-in. Then why is it showing in the diff?

shobhitagarwal1612 commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/EntityDao.java:
```diff
@@ -0,0 +1,45 @@
+/*
```
This class can be removed now

shobhitagarwal1612 commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/FeatureDao.java:
```diff
@@ -17,20 +17,14 @@
-  Completable insertOrUpdate(FeatureEntity feature);
-
+public interface FeatureDao extends BaseDao<FeatureEntity> {
   @Query("SELECT * FROM feature WHERE project_id = :projectId")
```
Something is fishy. These changes are already committed into the master branch.

shobhitagarwal1612 commented on gnd/src/main/java/com/google/android/gnd/persistence/local/room/RecordDao.java:
```diff
@@ -17,25 +17,13 @@
-
+public interface RecordDao extends BaseDao<RecordEntity> {
   /** Returns the observation with the specified UUID, if found. */
   @Query("SELECT * FROM record WHERE id = :recordId")
```
Same. Changes already checked in

shobhitagarwal1612 commented:
@gino-m Can you please update the branch? Some of the diff is showing already merged code.

gino-m commented:
@shobhitagarwal1612 Not sure why merge didn't work properly, but I've deleted the extra file and updated the Base in GitHub to force the diff to refresh, looks like it's working now.
